src/plugins/liveobjects/livemap.ts

```diff
@@ -53,18 +92,197 @@ export class LiveMap extends LiveObject<LiveMapData> {
+      return this._liveObjects.getPool().get(data.objectId);
     }
   }
 
   size(): number {
-    return this._dataRef.data.size;
+    let size = 0;
+    for (const value of this._dataRef.data.values()) {
+      if (value.tombstone === true) {
+        // should not count deleted entries
+        continue;
+      }
+
+      size++;
+    }
+
+    return size;
+  }
+
+  /**
+   * @internal
+   */
+  applyOperation(op: StateOperation, msg: StateMessage): void {
+    if (op.objectId !== this.getObjectId()) {
+      throw new this._client.ErrorInfo(
+        `Cannot apply state operation with objectId=${op.objectId}, to this LiveMap with objectId=${this.getObjectId()}`,
+        50000,
+        500,
+      );
+    }
+
+    switch (op.action) {
+      case StateOperationAction.MAP_CREATE:
+        if (this._client.Utils.isNil(op.map)) {
+          this._throwNoPayloadError(op);
+        } else {
+          this._applyMapCreate(op.map);
+        }
+        break;
+
+      case StateOperationAction.MAP_SET:
+        if (this._client.Utils.isNil(op.mapOp)) {
+          this._throwNoPayloadError(op);
+        } else {
+          this._applyMapSet(op.mapOp, msg.serial);
+        }
+        break;
+
+      case StateOperationAction.MAP_REMOVE:
+        if (this._client.Utils.isNil(op.mapOp)) {
+          this._throwNoPayloadError(op);
+        } else {
+          this._applyMapRemove(op.mapOp, msg.serial);
+        }
+        break;
+
+      default:
+        throw new this._client.ErrorInfo(
+          `Invalid ${op.action} op for LiveMap objectId=${this.getObjectId()}`,
+          50000,
+          500,
+        );
+    }
   }
 
   protected _getZeroValueData(): LiveMapData {
     return { data: new Map<string, MapEntry>() };
   }
+
+  private _throwNoPayloadError(op: StateOperation): void {
+    throw new this._client.ErrorInfo(
+      `No payload found for ${op.action} op for LiveMap objectId=${this.getObjectId()}`,
+      50000,
+      500,
+    );
+  }
+
+  private _applyMapCreate(op: StateMap): void {
+    if (this._semantics !== op.semantics) {
+      throw new this._client.ErrorInfo(
+        `Cannot apply MAP_CREATE op on LiveMap objectId=${this.getObjectId()}; map's semantics=${this._semantics}, but op expected ${op.semantics}`,
+        50000,
+        500,
+      );
+    }
+
+    // in order to apply MAP_CREATE op for an existing map, we should merge their underlying entries keys.
+    // we can do this by iterating over entries from MAP_CREATE op and apply changes on per-key basis as if we had MAP_SET, MAP_REMOVE operations.
+    Object.entries(op.entries ?? {}).forEach(([key, entry]) => {
+      // for MAP_CREATE op we must use dedicated timeserial field available on an entry, instead of a timeserial on a message
+      const opOriginTimeserial = entry.timeserial;
+      if (entry.tombstone === true) {
+        // entry in MAP_CREATE op is deleted, try to apply MAP_REMOVE op
+        this._applyMapRemove({ key }, opOriginTimeserial);
```

Comment:
this is interesting - although it shouldn't be allowed to initialise a map with tombstoned entries. Nothing theoretically stopping us, but not sure what the use case would be for that - did you have one in mind?

On the realtime side, we only handle the `set` case: https://github.com/ably/realtime/blob/main/go/realtime/lib/state/map.go#L115

If we decide to keep this, we should update the realtime side too.

Reply:
> what the use case would be for that - did you have one in mind?

not anything in particular, no. It just seemed natural and intuitive to handle this case in such a way, it aligns with everything just fine, and doesn't leave us in a situation "tombstone=true and we don't know what to do, let's throw an error"

Reply:
Good call - I've created a ticket to align the realtime side here: https://ably.atlassian.net/browse/DTP-1008